Module1/ansible/module4_playbook.yml

```diff
@@ -0,0 +1,178 @@
+---
+# YAML documents begin with the document separator --
+#
+# The minus in YAML indicates a list item.
+# The playbook contains a list of plays, with each play being a dictionary
+#
+
+- 
+  # Target: where our plays run and the options to run as
+  hosts: module4
+  gather_facts: false
+
+  vars_files:
+    - "secrets.yml"
+
+  vars:
+    packages:
+      - git
+      - python311
+
+    POETRY_USER: todoapp
+    POETRY_GROUP: ec2-user
+    POETRY_BIN: /home/{{ POETRY_USER }}/.local/bin/poetry
+
+  # Task: the list of tasks executed within the play
+  tasks:
+
+    # create the todoapp user
+    - name: create user todoapp
+      user:
+        name: "{{ POETRY_USER }}"
+        comment: APP user
+        shell: /bin/bash
+        group: "{{ POETRY_GROUP }}"
+      become: yes
+      register: useradd_output
+
+    - name: debug for user
+      debug:
+        var: useradd_output
+      tags: debug_output
+
+    # installation of packages
+    - name: package installation
+      yum:
+        name: "{{ packages }}"
+        state: latest
+      register: yum_install_output
+      become: yes
+
+    - name: debug
+      debug:
+        msg: "{{ yum_install_output }}"
+      tags: debug_output
+
+    # install poetry
+    - name: installation of poetry 
+      shell: |
+        curl -sSL https://install.python-poetry.org | python3 -
+      register: poetry_install_output
+      become: true
+      become_user: "{{ POETRY_USER }}"
+
+    - name: debug
+      debug:
+        msg: "{{ poetry_install_output }}"
+      tags: debug_output
+
+    # create directory
+    - name: create directory
+      file:
+        path: /opt/todoapp
+        owner: ec2-user
+        group: ec2-user
+        mode: '0755'
+        state: directory
+      register: dir_todoapp_output
+      become: yes
+
+    - name: dir output
+      debug:
+        msg: "{{ dir_todoapp_output }}"
+      tags: debug_output
+
+    # git
+    - name: Example clone of a single branch
+      git:
+        repo: https://github.com/harbymoo/DevOps-Course-Starter.git
+        dest: /opt/todoapp
+        single_branch: yes
+        version: main
+        update: no
+      register: git_output
+      notify: update permissions
+
+    - name: output git clone
+      debug:
+        msg: "{{ git_output }}"
+      tags: debug_output
+
+    # update the poetry
+    - name: poetry update
+      shell: |
+        {{ POETRY_BIN }} install --directory=/opt/todoapp/Module1
+      register: poetry_update_output
+      become: yes
+      become_user: "{{ POETRY_USER }}"
```

Comment:
Interestingly this step errors for me if I run it a second time with "Permission Denied"
> [Permission denied](fatal: [mod4node]: FAILED! => {"changed": true, "cmd": "/home/todoapp/.local/bin/poetry install --directory=/opt/todoapp/Module1\n", "delta": "0:00:01.348117", "end": "2023-11-14 09:51:30.551134", "msg": "non-zero return code", "rc": 1, "start": "2023-11-14 09:51:29.203017", "stderr": "\n[Errno 13] Permission denied: '/opt/todoapp/Module1/.venv/lib/python3.9/site-packages/todo_app.pth'", "stderr_lines": ["", "[Errno 13] Permission denied: '/opt/todoapp/Module1/.venv/lib/python3.9/site-packages/todo_app.pth'"], "stdout": "Installing dependencies from lock file\n\nNo dependencies to install or update\n\nInstalling the current project: todo-app (0.1.0)", "stdout_lines": ["Installing dependencies from lock file", "", "No dependencies to install or update", "", "Installing the current project: todo-app (0.1.0)"]})

There might be an issue with this not being idempotent somewhere